[JIT] Add JIT support for torch.no_grad

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41371 [JIT] Add JIT support for torch.no_grad**

**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

Commit:
[JIT] Add JIT support for torch.no_grad

**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

[ghstack-poisoned]

SplitInfinity commented:
Okay, so let me explain my thinking on this one.

After looking at the [source code for `torch.no_grad`](https://pytorch.org/docs/master/_modules/torch/autograd/grad_mode.html#no_grad) I concluded that it is not scriptable as is because it uses inheritance to allow `no_grad` to be used as a decorator, and some Python magic to correctly handle `no_grad` when it is applied to generators (which JIT doesn't support). The rest of the implementation, however, is quite simple:

```
    def __enter__(self):
        self.prev = torch.is_grad_enabled()
        torch._C.set_grad_enabled(False)

    def __exit__(self, *args):
        torch.set_grad_enabled(self.prev)
```

I think the best path forward is to create a JITable version of `no_grad` that has the same `__enter__` and `__exit__` implementations shown above, but without a dependency on Python. To this end, I added `aten::is_grad_enabled` and `aten::set_grad_enabled` that call `GradMode::is_enabled()` and `GradMode::set_enabled()` respectively. Thanks to preexisting name resolution mechanisms, `torch.is_grad_enabled` and `torch.set_grad_enabled` get emitted as the aforementioned ops. I wrote a new class called `NoGrad` in the scope of the tests, but I would like to hoist that out and put it somewhere that is globally discoverable and visible so that everyone can use it. Some additional magic will also be required to make sure that `torch.no_grad` resolves correctly and finds and uses that JITable `no_grad` implementation.

Some limitations of this approach:
- `no_grad` cannot be used as a decorator
- as I was writing this, I learned that `torch.set_grad_enabled` is itself something that can be used in a with statement. So, emitting it as an `aten::set_grad_enabled` op that calls `GradMode::set_enabled()` might have some unintended consequences if someone tries to JIT code that uses `torch.set_grad_enabled` directly.

SplitInfinity commented on test/jit/test_with.py:
```diff
@@ -548,3 +548,30 @@ def test_exit_incorrect_types(x, c):
+        Check that torch.no_grad() works.
+        """
+        @torch.jit.script
+        class NoGrad(object):
```
I put this class here temporarily, but I'd like to move it to a place where all users can access it and have `torch.no_grad` resolve to it somehow.

eellison commented on torch/csrc/jit/runtime/register_special_ops.cpp:
```diff
@@ -406,7 +406,14 @@ RegisterOperators reg({
+        aliasAnalysisFromSchema()),
+    Operator(
+        "aten::set_grad_enabled(bool val) -> ()",
+        [](Stack* stack) { torch::GradMode::set_enabled(pop(stack).toBool()); },
```
Pretty sure that returning `-> ()` in the schema will return an empty tuple. I think you have to push that or make it return None and return a None

SplitInfinity replied:
Oops, I thought this was the schema for returning nothing.

SplitInfinity replied:
Looks like `prim::Print` uses `-> ()` to denote no return value and pushes nothing to the stack.

eellison replied:
Yea, i'm not sure that's correct though. If you `print(torch.set_grad_enabled(True)` you'll get `None`, you could just return None here and add that to the stack.

SplitInfinity replied:
Done.

eellison reviewed (commented):
Looks like a good start, there's a couple things we have to figure out:

- it would be nice if we could figure out a way to support torch.no_grad() without having a special resolving mechanism for it. We should at least scope out what that would entail before deciding what to do. 

- how do we want to register operators that read/write from global state. Ideally this would be expressible in the schema, see https://github.com/pytorch/pytorch/pull/39497#issuecomment-640970375. A workaround is to use hasSideEffects.

eellison commented:
hey sorry i had this as an old tab and didn't see your comment on the PR, one sec re reading

eellison commented:
> I think the best path forward is to create a JITable version of no_grad that has the same __enter__ and __exit__ implementations shown above, but without a dependency on Python 

Is the main difficulty here is we have to make `torch.no_grad` work as a decorator as well? What about special-casing objects with _DecoratorContextManager as the one concrete supertype and ignoring the superclass ?

SplitInfinity commented:
> > I think the best path forward is to create a JITable version of no_grad that has the same **enter** and **exit** implementations shown above, but without a dependency on Python
> 
> Is the main difficulty here is we have to make `torch.no_grad` work as a decorator as well? What about special-casing objects with _DecoratorContextManager as the one concrete supertype and ignoring the superclass ?

Most of the JIT use cases I've seen are in `with` statements, not as decorators. That's also what I was focusing on in this PR. So I think the main problems are:
- dealing with the fact that `torch.no_grad` uses inheritance (your suggestion addresses this)
-making `torch.jit.no_grad()` resolve to a scripted version of the corresponding Python class in code that is JITed. In general, `torch.x` is assumed to be an `aten::x` op at the moment, and scripted classes have to be explicitly annotated with `@torch.jit.script` to be compiled and added to the compilation unit and show up during name resolution.

eellison commented:
> making torch.jit.no_grad() resolve to a scripted version of the corresponding Python class in code that is JITed

You can address this here: https://github.com/pytorch/pytorch/blob/master/torch/jit/_builtins.py#L118 It's how we resolve torch.unique and other ops bound in torch/functional.py _is_special_functional_bound_op

Commit:
Update on "[WIP][JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

[ghstack-poisoned]

SplitInfinity commented:
Okay, uploaded a new version after offline discussion with @eellison. This new version
- gets rid of the new `NoGrad` class I added and uses `torch.no_grad` directly
- makes some changes to `torch.no_grad` to make it scriptable; most of these seem harmless and I will add a reviewer for those changes once it is ready to review

The one problem left to solve is how to allow the implementation of `no_grad` to use inheritance and to ignore it for JIT purposes. This is usually not a good idea, but in this specific case, `no_grad` extends `_DecoratorContextManager` which allows it to be used as a decorator and on generator functions. These use cases are out of scope for JIT, so omitting the functionality of the base class in JIT would not be a problem.

Commit:
Update on "[WIP][JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

[ghstack-poisoned]

SplitInfinity commented on torch/jit/_builtins.py:
```diff
@@ -115,7 +115,7 @@ def _get_builtin_table():
         for name in dir(mod):
             v = getattr(mod, name)
-            if callable(v) and not _is_special_functional_bound_op(v):
+            if callable(v) and not _is_special_functional_bound_op(v) and "no_grad" not in name:
```
We should find a cleaner way to do this

eellison replied:
You should be able to check via identity here, similar to how https://github.com/pytorch/pytorch/blob/3ec5adf3360cbb721d54b6fec61a9e44b097c3e5/torch/jit/_builtins.py#L101 is directly comparing if the objects are the same

SplitInfinity replied:
Done.

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_with.py:
```diff
@@ -548,3 +548,18 @@ def test_exit_incorrect_types(x, c):
                 test_exit_incorrect_types, (test_tensor, ExitIncorrectTypes())
             )
+
+    def test_with_no_grad(self):
```
The only part of this that tests `no_grad` is `requires_grad`. Are there any numeric tests I can run here?

gchanan replied:
see below, you should look into what we test with no_grad when not scripting.

SplitInfinity commented on torch/autograd/grad_mode.py:
```diff
@@ -60,11 +61,16 @@ class no_grad(_DecoratorContextManager):
         >>> z.requires_grad
         False
     """
+    def __init__(self):
```
This might be useful to refactor out into a separate class.

SplitInfinity commented on torch/csrc/jit/runtime/register_special_ops.cpp:
```diff
@@ -406,7 +406,14 @@ RegisterOperators reg({
         aliasAnalysisFromSchema()),
-
+    Operator(
+        "aten::is_grad_enabled() -> bool",
```
Done.

SplitInfinity commented on torch/csrc/jit/runtime/register_special_ops.cpp:
```diff
@@ -406,7 +406,14 @@ RegisterOperators reg({
+        [](Stack* stack) { push(stack, torch::GradMode::is_enabled()); },
+        aliasAnalysisFromSchema()),
+    Operator(
+        "aten::set_grad_enabled(bool val) -> ()",
```
Done.

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]

gchanan commented on test/jit/test_with.py:
```diff
@@ -548,3 +548,18 @@ def test_exit_incorrect_types(x, c):
+
+            return y, y.requires_grad
+
+        test_input = torch.randn(5)
```
doesn't this test pass even if you don't have a no_grad block?

gchanan replied:
I don't know what the scope of support for autograd generally is with torchscript, but I think you'd want a subset of the no_grad tests in test_autograd, i.e.:
https://github.com/pytorch/pytorch/blob/34025eb8265b03ef83256a1a9c42f8876c9428f5/test/test_autograd.py#L912-L973
https://github.com/pytorch/pytorch/blob/34025eb8265b03ef83256a1a9c42f8876c9428f5/test/test_autograd.py#L2026-L2059
https://github.com/pytorch/pytorch/blob/34025eb8265b03ef83256a1a9c42f8876c9428f5/test/test_autograd.py#L4210-L4234

SplitInfinity replied:
Cool, I adapted a handful of these for JIT. I'll consult my JIT reviewers on whether what I added is enough.

gchanan commented on torch/autograd/grad_mode.py:
```diff
@@ -60,11 +61,16 @@ class no_grad(_DecoratorContextManager):
+        torch.set_grad_enabled(False)
 
-    def __exit__(self, *args):
+    def __exit__(self, exc: Any, type: Any, tb: Any):
```
probably this doesn't matter, but it's documented as taking `exc_type, exc_value, traceback`.

SplitInfinity replied:
Good catch, I changed this.

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]

SplitInfinity commented on torch/autograd/grad_mode.py:
```diff
@@ -60,11 +61,16 @@ class no_grad(_DecoratorContextManager):
         False
     """
+    def __init__(self):
+        if not torch._jit_internal.is_scripting():
+            super().__init__()
+        self.prev = False
+
     def __enter__(self):
         self.prev = torch.is_grad_enabled()
-        torch._C.set_grad_enabled(False)
+        torch.set_grad_enabled(False)
```
@gchanan 

What do you think of these changes? In particular, what are the implications of changing `torch._C.set_grad_enabled` to `torch.set_grad_enabled`?

eellison replied:
These are fine, it's equivalent

albanD replied:
Actually this is the context manager below ;) that then calls the `_C` version.

SplitInfinity commented:
Ping @albanD @eellison

eellison commented on torch/autograd/grad_mode.py:
```diff
@@ -60,11 +61,16 @@ class no_grad(_DecoratorContextManager):
+    def __init__(self):
+        if not torch._jit_internal.is_scripting():
+            super().__init__()
+        self.prev = False
```
we shouldn't be setting `prev = False` for all `_DecoratorContextManager`

SplitInfinity replied:
Isn't this `no_grad.__init__`, not `_DecoratorContextManager.__init__`? I think the point still stands though. Is it better to use `Optional[bool]` here and refine later in `__exit__`?

eellison replied:
Oh yea, you're right. The way git displays this really looks look its `_DecoratorContextManager` lol. Yea this is fine as is, no need to use Optional.

eellison reviewed (commented):
Looks good, a small comments.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -12332,7 +12332,7 @@ def close_match(x):
             @torch.jit.script
             def unknown_op(x):
-                torch.set_grad_enabled(True)
+                torch.set_anomaly_enabled(True)
```
Because `set_grad_enabled` is now supported, but `set_anomaly_enabled` is not. This test is for some unknown op.

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM
You can rebase on top of master and it should fix the windows CI.

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add JIT support for torch.no_grad"


**Summary**
This commit enables the use of `torch.no_grad()` in a with item of a
with statement within JIT. Note that the use of this context manager as
a decorator is not supported.

**Test Plan**
This commit adds a test case to the existing with statements tests for
`torch.no_grad()`.

**Fixes**
This commit fixes #40259.

Differential Revision: [D22649519](https://our.internmc.facebook.com/intern/diff/D22649519)

[ghstack-poisoned]
